internal/cmd/users.go

```diff
@@ -225,51 +227,68 @@ func updateUser(cli *CLI, name string) error {
 		return err
 	}
 
-	req := &api.UpdateUserRequest{}
+	config, err := currentHostConfig()
+	if err != nil {
+		return err
+	}
 
 	isSelf, err := isUserSelf(name)
 	if err != nil {
 		return err
 	}
 
 	if isSelf {
-		config, err := currentHostConfig()
-		if err != nil {
+		req := &api.UpdateUserRequest{}
+		if req.ID, err = config.UserID.ID(); err != nil {
 			return err
 		}
 
-		if req.ID, err = config.PolymorphicID.ID(); err != nil {
+		fmt.Fprintf(cli.Stderr, "  Enter a new password (min. length 8):\n")
+		req.Password, err = promptSetPassword(cli, "")
+		if err != nil {
 			return err
 		}
-	} else {
-		user, err := getUserByName(client, name)
-		if err != nil {
-			if errors.Is(err, ErrUserNotFound) {
-				return Error{Message: fmt.Sprintf("No user named %q in local provider; only local users can be edited", name)}
-			}
+
+		if _, err := client.UpdateUser(req); err != nil {
 			return err
 		}
 
-		req.ID = user.ID
+		cli.Output("  Updated password")
+
+		return nil
 	}
 
-	fmt.Fprintf(cli.Stderr, "  Enter a new password (min. length 8):\n")
-	req.Password, err = promptSetPassword(cli, "")
+	ok, err := hasAccessToChangePasswordsForOtherUsers(client, config)
 	if err != nil {
 		return err
 	}
+	if !ok {
+		return Error{Message: "No permission to change password for user " + name}
+	}
 
-	logging.S.Debugf("call server: update user %s", req.ID)
-	if _, err := client.UpdateUser(req); err != nil {
+	user, err := getUserByName(client, name)
+	if err != nil {
+		if errors.Is(err, ErrUserNotFound) {
+			logging.S.Debugf("user not found: %s", err)
+			return Error{Message: fmt.Sprintf("No user named %q", name)}
+		}
 		return err
 	}
 
-	if isSelf {
-		cli.Output("  Updated password")
-	} else {
-		cli.Output("  Updated password for %q", name)
+	tmpPassword, err := generate.CryptoRandom(12, generate.CharsetPassword)
+	if err != nil {
+		return err
 	}
 
+	if _, err := client.UpdateUser(&api.UpdateUserRequest{
```

Comment:
Depends: 
1. There is an Infra user for this email: The API call assumes you mean that Infra user (since you can't update non-infra users)
2. There is not an infra user for this email: The API call fails. You can't update a non-local non-infra user. One option is to make this work by assuming you want to create the local infra user, but it seems really weird to do that. I'd rather have them explicitly create the local infra user first, then try to edit it.